bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -747,13 +749,13 @@ ClientRequestMsg *ReplicaImp::addRequestToPrePrepareMessage(ClientRequestMsg *&n
 // second value in the returned pair.
 // The first value of the pair can be nullptr depending on the availability of the result. So the
 // first value cannot represent failure.
-std::pair<PrePrepareMsg *, bool> ReplicaImp::finishAddingRequestsToPrePrepareMsg(PrePrepareMsg *&prePrepareMsg,
-                                                                                 uint32_t maxSpaceForReqs,
-                                                                                 uint32_t requiredRequestsSize,
-                                                                                 uint32_t requiredRequestsNum) {
+std::pair<std::unique_ptr<PrePrepareMsg>, bool> ReplicaImp::finishAddingRequestsToPrePrepareMsg(
+    std::unique_ptr<PrePrepareMsg> prePrepareMsg,
```

Comment:
long mesage here. in short: check deeply what is the prupose of passing *&prePrepareMsg in original code. OR maybe you already checked and I've just wrote all this for nothing, but I prefer to double check.

original code and (maybe) current look incorrect.
It might work, but in the general definition it's incorrect.
caller pass a reference to a pointer.
means if pointer is deleted, it should be set into null, else caller will have a dnagling pointer, which is very bad. so orignal code should have set a null after all delertes, but it didnt.

Did you mean to change how function is called or you didnt understand the reference to a pointer? i am not looking at caller code now, so I can't tell.
but if you want to keep the exact same pattern, you should pass a reference to unique_ptr (very ugly, but that's what we have for now). An alternative is to change how call is made and for that need to check code in all callers sides. 
An alternativwe would be to consider using here shared_ptr.

Might this be the reason for your memory corruption? (wild specualtion)

Reply:
Hi. The unique_ptr here solves all these issues since it must be moved in the caller.
So everybody knows that we shouldn't use the pointer after it was moved (and if someone, by mistake, tries to - clang tidy will catch it).